app/controllers/restaurants_controller.rb

```diff
@@ -1,16 +1,62 @@
+require_relative './application_controller'
+require 'active_record'
+
 class RestaurantsController < ApplicationController
-  def list
+  before_action :set_restaurant, only: [:show, :update, :destroy]
+  #TODO: consider adding authentication
+  rescue_from ActiveRecord::RecordNotFound, :with => :invalid_id_handler
+
+  # GET /restaurants
+  def index
+    @restaurants = Restaurant.all
+    @restaurants = @restaurants.where('cuisine = ?', query_params["by_cuisine"]) if query_params.has_key? "by_cuisine"
+    @restaurants = @restaurants.where('rating >= ?', query_params["min_rating"].to_i) if query_params.has_key? "min_rating"
+    @restaurants = @restaurants.where('delivery_time <= ?', query_params["max_delivery_time"].to_i) if query_params.has_key? "max_delivery_time"
+    render :json => @restaurants
   end
 
-  def retrieve
+  # GET /restaurants/:id
+  def show
+    @restaurant
+    render :json => @restaurant
   end
 
+  #POST /restaurants
   def create
+    @restaurant = Restaurant.create(restaurant_params)
+    #TODO: could there be other errors other than bad request (failed model validations) from this object?
+    @restaurant.valid? ? (render :json => @restaurant, :status => :created) : (render :json => @restaurant.errors.full_messages, :status => :bad_request)
   end
 
+  #PATCH /restaurants:/:id
   def update
+    @restaurant.update(restaurant_params)
```

Comment:
Same comment goes here with regards to the DB function, you can use `update!` and then skip the `valid?` line